src/pb_convert.rs

```diff
@@ -12,338 +12,505 @@
+                match pb.#oneof {
+                    #( #match_arms )*
+                    None => Err(failure::format_err!("Failed to decode #name from protobuf"))
+                }
             }
-        }
+        };
+        let to_pb_impl = {
+            let match_arms = self.variants.iter().map(|variant| {
+                let pb_variant = self.get_variant_name(variant);
+                let variant_name = &variant.name;
+
+                let setter = Ident::new(&format!("set_{}", pb_variant), Span::call_site());
+                quote! {
+                    #name::#variant_name(msg) => inner.#setter(msg.to_pb()),
+                }
+            });
 
-        impl<'de> _serde::Deserialize<'de> for #name {
-            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-            where
-                D: _serde::Deserializer<'de>,
-            {
-                let pb = <#name as ProtobufConvert>::ProtoStruct::deserialize(deserializer)?;
-                ProtobufConvert::from_pb(pb).map_err(_serde::de::Error::custom)
+            quote! {
+                let mut inner = Self::ProtoStruct::new();
+                match self {
+                    #( #match_arms )*
+                }
+                inner
+            }
+        };
+
+        quote! {
+            impl ProtobufConvert for #name {
+                type ProtoStruct = #pb_name;
+
+                fn from_pb(mut pb: Self::ProtoStruct) -> std::result::Result<Self, failure::Error> {
+                    #from_pb_impl
+                }
+
+                fn to_pb(&self) -> Self::ProtoStruct {
+                    #to_pb_impl
+                }
             }
         }
     }
-}
 
-pub fn implement_protobuf_convert(input: TokenStream) -> TokenStream {
-    let input: DeriveInput = syn::parse(input).unwrap();
+    fn impl_enum_conversions(&self) -> impl ToTokens {
+        let name = &self.name;
+
+        if self.attrs.impl_from_trait {
+            let conversions = self.variants.iter().map(|variant| {
+                let variant_name = &variant.name;
+                let field_name = &variant.field_name;
+                quote! {
+                    impl From<#field_name> for #name {
+                       fn from(variant: #field_name) -> Self {
+                           #name::#variant_name(variant)
+                       }
+                    }
 
-    let name = input.ident.clone();
-    let proto_struct_name = get_protobuf_struct_path(&input.attrs);
+                    impl std::convert::TryFrom<#name> for #field_name {
+                        type Error = failure::Error;
+
+                        fn try_from(msg: #name) -> Result<Self, Self::Error> {
+                            if let #name::#variant_name(inner) = msg {
+                                Ok(inner)
+                            } else {
+                                Err(failure::format_err!(
+                                    "Expected #variant_name variant, but got {:?}",
```

Comment:
`#variant_name` will not be escaped here, see

https://docs.rs/quote/1.0.2/quote/macro.quote.html#interpolating-text-inside-of-doc-comments

And it would be nice to have a test for the impl_from_trait feature, a simple conversion between `EnumMessage` and `Message` would be enough.